variables.tf

```diff
@@ -0,0 +1,59 @@
+variable "config" {
+  type = object({
+    server_type = optional(string, "cx11")
+    datacenter  = optional(string, "fsn1-dc14")
+
+    sshfs = object({
+      ip   = string
+      user = string
+      path = string
+    })
+
+    network_id     = optional(string, null)
+    network_cidr   = optional(string, "10.0.0.0/24")
+    network_subnet = optional(string, "10.0.0.0/28")
+  })
+
+  description = <<-EOF
+    server_type: VM type of the Minio server
+    datacenter: Datacenter of the Minio server
+
+    sshfs: SSHFS Server configuration
+    ip: IP address of the SSHFS Server
+    user: Username of the SSHFS Server
+    path: Path to data on the SSHFS Server
+
+    network_id: ID of the network
+    network_cidr: CIDR of the network
+    network_subnet: Subnet of the network
+    EOF
+}
+
+variable "sshfs_ssh_key" {
+  description = "SSH key for SSHFS Server"
+  type        = string
+  sensitive   = true
+}
+
+variable "minio_username" {
+  description = "Username for Minio root user"
+  type        = string
+
+  default = null
+
+  validation {
+    error_message = "Minio username must be at least 3 characters long."
+    condition     = var.minio_username != null ? length(var.minio_username) >= 3 : true
+  }
+}
+
+variable "minio_password" {
+  description = "Password for Minio root user"
+  type        = string
+  default     = null
+
+  validation {
+    error_message = "Minio password must be at least 8 characters long."
+    condition     = var.minio_password != null ? length(var.minio_password) >= 8 : true
+  }
+}
```

Comment:
There is a certain lack of consistency in variables. Some are grouped in objects, some are just individual variables. I would advice sticking to just plain variables, or be consistent with grouping (see suggestion).  
Also, _config_ is a bit meta. All variables are configuration.
```suggestion
variable "network" {
  type = object({
    id     = optional(string, null)
    cidr   = optional(string, "10.0.0.0/24")
    subnet = optional(string, "10.0.0.0/28")
  })

  description = <<-EOF
    id: ID of a pre-provisioned network. Optional, if not set a new network will be created.
    cidr: CIDR of the network
    subnet: Subnet of the network
    EOF
}

variable "server" {
    type = object({
      server_type = optional(string, "cx11")
      datacenter  = optional(string, "fsn1-dc14")
    })
    description = <<-EOF
    server_type: VM type (cpu architecture and size) of the Minio server
    datacenter: Datacenter of the Minio server
    EOF
}

variable "sshfs" {
    sshfs = object({
      ip          = string
      user        = string
      path        = string
      private_key = string
    })
    description = <<-EOF
    sshfs: SSHFS Server configuration
    ip: IP address of the SSHFS Server
    user: Username of the SSHFS Server
    path: Path to data on the SSHFS Server
    private_key: Private key for authentication with SSHFS Server
    EOF
}

variable "minio" {
    type = object({
      username = optional(string, null)
      password = optional(string, null)
    })
    description = <<-EOF
    username: Username for Minio root user
    password: Password for Minio root user
    EOF

    validation {
    error_message = "Minio username must be at least 3 characters long."
    condition     = var.minio.username != null ? length(var.minio.username) >= 3 : true
    }

    validation {
      error_message = "Minio password must be at least 8 characters long."
      condition     = var.minio.password != null ? length(var.minio.password) >= 8 : true
    }
}
```

Reply:
right, so I think using objects as input variables fit a few purposes:
1. The one you describe with input validation
2. prefixing variable names with eg. `sshfs_`, `minio_` and so on _can_ sometimes feel a bit silly. It is not too bad in this case, but sometimes prefixing gets quite long and trivial. In those cases objects feel a bit slimmer. Sort of like namespacing. Obviously a matter of taste.
3. If you pass the same (huge) configuration down a series of modules, repeating the same variables over and over gets very trivial. I think this is why many adopted the pattern in the first place.

However, if none of the 3 purposes above applies, I will always prefer individual (simple) input variables. If nothing else, it makes documentation much much cleaner